src/graph/ford_fulkerson.rs

```diff
@@ -1,141 +1,324 @@
+        ),
+        test_improper_graph: (
+            vec![
+                vec![0, 8],
+                vec![0],
+            ],
+            0,
+            1,
+            Err(FordFulkersonError::ImproperGraph),
+        ),
+        test_graph_with_small_flow: (
+            vec![
+                vec![0, 8, 0, 0, 3, 0],
+                vec![0, 0, 9, 0, 0, 0],
+                vec![0, 0, 0, 0, 7, 2],
+                vec![0, 0, 0, 0, 0, 5],
+                vec![0, 0, 7, 4, 0, 0],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            0,
+            5,
+            Ok(6),
+        ),
+        test_graph_with_medium_flow: (
+            vec![
+                vec![0, 10, 0, 10, 0, 0],
+                vec![0, 0, 4, 2, 8, 0],
+                vec![0, 0, 0, 0, 0, 10],
+                vec![0, 0, 0, 0, 9, 0],
+                vec![0, 0, 6, 0, 0, 10],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            0,
+            5,
+            Ok(19),
+        ),
+        test_graph_with_large_flow: (
+            vec![
+                vec![0, 12, 0, 13, 0, 0],
+                vec![0, 0, 10, 0, 0, 0],
+                vec![0, 0, 0, 13, 3, 15],
+                vec![0, 0, 7, 0, 15, 0],
+                vec![0, 0, 6, 0, 0, 17],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            0,
+            5,
+            Ok(23),
+        ),
+        test_complex_graph: (
+            vec![
+                vec![0, 16, 13, 0, 0, 0],
+                vec![0, 0, 10, 12, 0, 0],
+                vec![0, 4, 0, 0, 14, 0],
+                vec![0, 0, 9, 0, 0, 20],
+                vec![0, 0, 0, 7, 0, 4],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            0,
+            5,
+            Ok(23),
+        ),
+        test_disconnected_graph: (
+            vec![
+                vec![0, 0, 0, 0],
+                vec![0, 0, 0, 1],
+                vec![0, 0, 0, 1],
+                vec![0, 0, 0, 0],
+            ],
+            0,
+            3,
+            Ok(0),
+        ),
+        test_unconnected_sink: (
+            vec![
+                vec![0, 4, 0, 3, 0, 0],
+                vec![0, 0, 4, 0, 8, 0],
+                vec![0, 0, 0, 3, 0, 2],
+                vec![0, 0, 0, 0, 6, 0],
+                vec![0, 0, 6, 0, 0, 6],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            0,
+            5,
+            Ok(7),
+        ),
+        test_no_edges: (
+            vec![
+                vec![0, 0, 0],
+                vec![0, 0, 0],
+                vec![0, 0, 0],
+            ],
+            0,
+            2,
+            Ok(0),
+        ),
+        test_single_vertex: (
+            vec![
+                vec![0],
+            ],
+            0,
+            0,
+            Ok(0),
+        ),
+        test_negative_capacity: (
+            vec![
+                vec![0, -10, 0],
+                vec![0, 0, -5],
+                vec![0, 0, 0],
+            ],
```

Comment:
This is something what bothers me: do we really want to have inputs with negative entries?

Reply:
Because the graph is `Vec<Vec<isize>>`, I think it's not a redundant test. What do you think about it?

Reply:
If you feel this test is not necessary, feel free to remove it. I added this test because the implementation should handle the cases where the graph has negative flows.

Reply:
First of all, I am not sure if negative numbers make any sense here. Secondly, please have a look at this line:
https://github.com/TheAlgorithms/Rust/blob/2576c211e200deb8f1998a71a1f32dd71608d9cb/src/graph/ford_fulkerson.rs#L40

What should be the output for
```rust
vec![
    vec![0, -1, 1],
    vec![-1, 0, 100],
    vec![1, 100, 0],
],
```
99 or 1?

Reply:
Yes, I overlook it. The negative flows do not make any sense in this implementation.

Reply:
Because the flow always is non-negative values, we can make the following changes:
- The graph representation should be `Vec<Vec<usize>`
- The parent array in the Ford-Fulkerson algorithm is used to store the augmenting path found by the BFS and should be an array of `usize` values (eliminate type casting from `isize` to `usize`)
- The tests for negative cases should be removed (it does not make any sense in the implementation)

Reply:
Let's see the change @vil02?